README.md
# Unfurl Links

Unfurl links on Issues and Pull Request discussions.

![screenshot](https://github.com/wow-actions/unfurl-links/blob/master/screenshots/default.jpg?raw=true)

## Usage

Create `.github/workflows/unfurl-links.yml` in the default branch:

```yml
name: Unfurl Links
on:
  issues:
    types: [opened, edited]
  issue_comment:
    types: [created, edited]
  pull_request:
    types: [opened, edited]
jobs:
  run:
    runs-on: ubuntu-latest
    steps:
      - uses: wow-actions/unfurl-links@v1
        with:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```

### Inputs

#### GITHUB_TOKEN

Your GitHub token for authentication.

#### raw

Specify if only render the raw links. Default `true`.

![raw links](https://github.com/wow-actions/unfurl-links/blob/master/screenshots/raw-links.jpg?raw=true)

Set `raw` to `false` to render all links.

```yml
name: Unfurl Links
on:
  issues:
    types: [opened, edited]
  issue_comment:
    types: [created, edited]
  pull_request:
    types: [opened, edited]
jobs:
  run:
    runs-on: ubuntu-latest
    steps:
      - uses: wow-actions/unfurl-links@v1
        with:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
          raw: false
```

![all links](https://github.com/wow-actions/unfurl-links/blob/master/screenshots/all-links.jpg?raw=true)

#### header

Cutsom header of the unfurled card. `header` can be a [Handlebars](https://handlebarsjs.com/) template and rendered with parsed metadata.

```yml
name: Unfurl Links
on:
  issues:
    types: [opened, edited]
  issue_comment:
    types: [created, edited]
  pull_request:
    types: [opened, edited]
jobs:
  run:
    runs-on: ubuntu-latest
    steps:
      - uses: wow-actions/unfurl-links@v1
        with:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
          header: '<i><a href="{{ url }}">{{ url }}</a></i>'
```

![custom header](https://github.com/wow-actions/unfurl-links/blob/master/screenshots/custom-header.jpg?raw=true)

#### template

[Handlebars](https://handlebarsjs.com/) template to render the unfurled card. The template will be rendered with parsed metadata from url.

```ts
interface Metadata {
  url: string
  header?: string
  title?: string
  titleLink?: string
  authorName?: string
  authorIcon?: string
  authorLink?: string
  thumb?: string
  content?: string
  image?: string
  footer?: string
  footerLink?: string
  footerIcon?: string
}
```

And the default template is:

```hbs
<blockquote>
  {{#if header}}
    {{{header}}}
  {{/if}}

  {{#if thumb}}
    <img src='{{thumb}}' width='48' align='right' />
  {{/if}}

  {{#if authorName}}
    <div>
      {{#if authorIcon}}
        <img src='{{authorIcon}}' height='14' />
      {{/if}}
      {{#if authorLink}}
        <a href='{{authorLink}}'>{{authorName}}</a>
      {{else}}
        {{authorName}}
      {{/if}}
    </div>
  {{/if}}

  {{#if title}}
    <div>
      <strong>
        {{#if titleLink}}
          <a href='{{titleLink}}'>{{title}}</a>
        {{else}}
          {{title}}
        {{/if}}
      </strong>
    </div>
  {{/if}}

  {{#if content}}
    <div>{{content}}</div>
  {{/if}}

  {{#if image}}
    <br />
    <img src='{{image}}' />
  {{/if}}

  {{#if footer}}
    <h6>
      {{#if footerIcon}}
        <img src='{{footerIcon}}' height='14' />
      {{/if}}
      {{#if footerLink}}
        <a href='{{footerLink}}'>{{footer}}</a>
      {{else}}
        {{footer}}
      {{/if}}
    </h6>
  {{/if}}

</blockquote>
```

## License

The scripts and documentation in this project are released under the [MIT License](LICENSE)


CHANGELOG.md
## [1.0.1](https://github.com/wow-actions/unfurl-links/compare/v1.0.0...v1.0.1) (2021-09-20)


### Bug Fixes

* typos ([254f2d7](https://github.com/wow-actions/unfurl-links/commit/254f2d7c6d51e5cda6254af142a12ea5351f288b))

# 1.0.0 (2021-09-18)


### Bug Fixes

* 🐛 image url ([438f96c](https://github.com/wow-actions/unfurl-links/commit/438f96c44e413211504cc6095439a646cb016d4f))
* 🐛 regex of attachment ([959608b](https://github.com/wow-actions/unfurl-links/commit/959608b22bfbdc87044f5e63039353b01b1d0203))
* 🐛 regex to match old unfurl ([df2c58e](https://github.com/wow-actions/unfurl-links/commit/df2c58e8c62dea5f16712fee44dac713b9efdf36))
* 🐛 remove line breaks in template ([6d3b270](https://github.com/wow-actions/unfurl-links/commit/6d3b270be8810b303111667ffc43a66a40bcbfa9))
* 🐛 replace unfurls when update comment ([a94359e](https://github.com/wow-actions/unfurl-links/commit/a94359ea2c68d3c632b004cb641c157a58a88014))
* 🐛 template ([034edd7](https://github.com/wow-actions/unfurl-links/commit/034edd7849b887c14ea61e1db49899045373df98))
* 🐛 trigger action on comment cteated ([4ecea37](https://github.com/wow-actions/unfurl-links/commit/4ecea37070292d4ac8142c47129ee6e7e954f631))


### Features

* ✨ custom template ([e79760a](https://github.com/wow-actions/unfurl-links/commit/e79760a29696092d729220ca99a80e41bce0ed72))
* ✨ init ([84b7ffc](https://github.com/wow-actions/unfurl-links/commit/84b7ffce62fc080dbe569a90ab119a84437aa432))
* ✨ parse all links ([6a5f4ef](https://github.com/wow-actions/unfurl-links/commit/6a5f4ef0bcbedba9b48aa13b3a78f8d386988179))
* ✨ thumb option ([c5f50a3](https://github.com/wow-actions/unfurl-links/commit/c5f50a3410d9cf730952c047be20591df7edd792))


src/action.ts
import * as core from '@actions/core'
import * as github from '@actions/github'
import { Util } from './util'
import { Unfurl } from './unfurl'

export namespace Action {
  export async function run() {
    try {
      if (
        Util.isValidEvent('pull_request', 'opened') ||
        Util.isValidEvent('pull_request', 'edited') ||
        Util.isValidEvent('issues', 'opened') ||
        Util.isValidEvent('issues', 'edited') ||
        Util.isValidEvent('issue_comment', 'created') ||
        Util.isValidEvent('issue_comment', 'edited')
      ) {
        const { context } = github
        const headers = { accept: 'application/vnd.github.full+json' }

        if (context.payload.comment) {
          const octokit = Util.getOctokit()
          const comment = await octokit.rest.issues.getComment({
            ...context.repo,
            headers,
            comment_id: context.payload.comment.id,
          })
          // core.debug(JSON.stringify(comment.data, null, 2))
          const html = Util.getHtml(comment.data)
          const body = await Unfurl.parse(html)
          if (body) {
            await octokit.rest.issues.updateComment({
              ...context.repo,
              body,
              comment_id: context.payload.comment.id,
            })
          }
        } else {
          const payload = context.payload.pull_request || context.payload.issue
          if (payload) {
            if (context.payload.pull_request) {
              if (payload.draft) {
                core.debug('Ignore draft PR')
                return
              }
            }

            const octokit = Util.getOctokit()
            const issue = await octokit.rest.issues.get({
              ...context.repo,
              headers,
              issue_number: payload.number,
            })

            const html = Util.getHtml(issue.data)
            const body = await Unfurl.parse(html)
            if (body) {
              await octokit.rest.issues.update({
                ...context.repo,
                body,
                issue_number: payload.number,
              })
            }
          }
        }
      }
    } catch (e) {
      core.error(e)
      core.setFailed(e.message)
    }
  }
}


src/template.ts
export namespace Template {
  export interface Metadata {
    pretext?: string
    title?: string
    titleLink?: string
    authorName?: string
    authorIcon?: string
    authorLink?: string
    thumbUrl?: string
    content?: string
    imageUrl?: string
    footer?: string
    footerLink?: string
    footerIcon?: string
  }

  // https://github.com/probot/attachments/blob/master/template.js
  export function render(data: Metadata) {
    let str = ''

    if (data.pretext) {
      str += `${data.pretext}\n`
    }

    str += '<blockquote>'

    if (data.thumbUrl) {
      str += `<img src="${data.thumbUrl}" width="48" align="right">`
    }

    if (data.authorName) {
      str += '<div>'

      if (data.authorIcon) {
        str += `<img src="${data.authorIcon}" height="14"> `
      }

      if (data.authorLink) {
        str += `<a href="${data.authorLink}">${data.authorName}</a>`
      } else {
        str += data.authorName
      }

      str += '</div>'
    }

    if (data.title) {
      str += `<div><strong>`

      if (data.titleLink) {
        str += `<a href="${data.titleLink}">${data.title}</a>`
      } else {
        str += data.title
      }

      str += `</strong></div>`
    }

    if (data.content) {
      str += `<div>${data.content}</div>`
    }

    if (data.imageUrl) {
      str += `<br><img src="${data.imageUrl}">`
    }

    if (data.footer) {
      str += '<h6>'

      if (data.footerIcon) {
        str += `<img src="${data.footerIcon}" height="14"> `
      }

      if (data.footerLink) {
        str += `<a href="${data.footerLink}">${data.footer}</a>`
      } else {
        str += data.footer
      }

      str += '</h6>'
    }

    str += '</blockquote>'

    return str
  }
}


src/unfurl.ts
import * as core from '@actions/core'
import { JSDOM } from 'jsdom'
import { unfurl } from 'unfurl.js'
import { compile } from 'handlebars'

export namespace Unfurl {
  // eslint-disable-next-line no-inner-declarations
  function getLinks(html: string) {
    const links: string[] = []

    // Parse the HTML
    const frag = JSDOM.fragment(html)
    const raw = core.getInput('raw') !== 'false'

    // Check if the content came from an email
    const isEmail = !!frag.querySelectorAll('.email-fragment').length

    if (!isEmail) {
      // Find all the links
      frag.querySelectorAll('a').forEach((a) => {
        // Only unfurl raw links. GitHub also adds a trailing `/`
        if (!raw || a.href === a.innerHTML || a.href === `${a.innerHTML}/`) {
          links.push(a.href)
        }
      })
    }

    // Avoid duplicate identical links
    return Array.from(new Set(links))
  }

  interface Metadata {
    url: string
    header?: string
    title?: string
    titleLink?: string
    authorName?: string
    authorIcon?: string
    authorLink?: string
    thumb?: string
    content?: string
    image?: string
    footer?: string
    footerLink?: string
    footerIcon?: string
  }

  // eslint-disable-next-line no-inner-declarations
  async function getMetadata(url: string): Promise<Metadata> {
    const data = await unfurl(url, { oembed: true })
    const ogps = data.open_graph
    const cards = data.twitter_card
    const header = core.getInput('header')

    const ogp = Array.isArray(ogps) ? ogps[0] : ogps
    const card = Array.isArray(cards) ? cards[0] : cards
    const embed = data.oEmbed

    const content =
      (ogp && ogp.description) || (card && card.description) || data.description

    const title = (ogp && ogp.title) || (card && card.title) || data.title
    const titleLink = (ogp && ogp.url) || url

    const authorName = embed && embed.author_name
    const authorLink = embed && embed.author_url

    const thumb =
      (ogp && ogp.images && ogp.images[0] && ogp.images[0].url) ||
      (card && card.images && card.images[0] && card.images[0].url) ||
      (embed &&
        embed.thumbnails &&
        embed.thumbnails[0] &&
        embed.thumbnails[0].url)

    return {
      url,
      header,
      title,
      titleLink,
      authorName,
      authorLink,
      content,
      thumb,
    }
  }

  const defaultTemplate = `
    <blockquote>
      {{#if header }}
        {{{ header }}}
      {{/if}}

      {{#if thumb }}
        <img src="{{ thumb }}" width="48" align="right" />
      {{/if}}

      {{#if authorName }}
        <div>
          {{#if authorIcon }}
            <img src="{{ authorIcon }}" height="14" />
          {{/if}}
          {{#if authorLink }}
            <a href="{{ authorLink }}">{{ authorName }}</a>
          {{else}}
            {{ authorName }}
          {{/if}}
        </div>
      {{/if}}

      {{#if title }}
        <div>
          <strong>
            {{#if titleLink }}
              <a href="{{ titleLink }}">{{ title }}</a>
            {{else}}
              {{ title }}
            {{/if}}
          </strong>
        </div>
      {{/if}}

      {{#if content }}
        <div>{{ content }}</div>
      {{/if}}

      {{#if image }}
        <br/>
        <img src="{{image}}" />
      {{/if}}

      {{#if footer }}
        <h6>
          {{#if footerIcon }}
            <img src="{{ footerIcon }}" height="14" />
            {{/if}}
          {{#if footerLink }}
            <a href="{{ footerLink }}">{{ footer }}</a>
          {{else}}
            {{ footer }}
          {{/if}}
        </h6>
      {{/if}}

    </blockquote>
  `

  export async function parse(html: string) {
    // https://regex101.com/r/m6GyIi/1
    const regex =
      /[\r\n\s]*<!--\s*unfurl\s+begin\s*-->([\S\s]*?)[\r\n\s]*<!--\s*unfurl\s+end\s*-->[\r\n\s]*/gm
    const raw = html.replace(regex, '')
    const links = getLinks(raw)

    core.debug(`html: ${html}`)
    core.debug(`raw: ${raw}`)
    core.debug(`links: ${links}`)

    if (links.length) {
      const template = core.getInput('template') || defaultTemplate
      const render = compile(template.replace(/[\t\r\n]*/g, ''))

      const contents = await Promise.all(
        links.map((link) =>
          getMetadata(link).then((data) => {
            if (data.header) {
              data.header = compile(data.header)(data)
            }
            return render(data).trim()
          }),
        ),
      )

      const prefix = '<!-- unfurl begin -->'
      const suffix = '<!-- unfurl end -->'

      return `${raw}\n\n${prefix}\n\n${contents.join('')}\n\n${suffix}`
    }

    return null
  }
}
